make env optional arg while creating from buffers

avjmachine opened the pull request:
# Description

Make the env argument as optional in the create_dataset_from_buffers() function in utils

Fixes # (issue), Depends on # (pull request)

## Type of change

> Please delete options that are not relevant.

- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

### Screenshots

> Please attach before and after screenshots of the change if applicable.
> To upload images to a PR -- simply drag and drop or copy paste.

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have run `pytest -v` and no errors are present.
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I solved any possible warnings that `pytest -v` has generated that are related to my code to the best of my knowledge.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

balisujohn reviewed (changes requested):
Thanks for your work on this so far. I requested changes in two places. I also think there should be a new test which creates a dataset without providing an `env` argument, and makes sure that everything still works as intended. Once that is all done, I will review again.

balisujohn commented:
Thanks for your work on this; it looks almost ready! I think it would be good to add a test to test_dataset_creation where a dataset is created from buffers without the env being set, and then also tries to load that dataset after it is created. Additionally, it looks like pre-commit is failing. Once those two things are done, it should be ready to merge.

younik commented on minari/utils.py:
```diff
@@ -433,10 +433,16 @@ def create_dataset_from_buffers(
-        observation_space = env.observation_space
-    if action_space is None:
-        action_space = env.action_space
+    if env is None:
+        if observation_space is None or action_space is None:
+            raise ValueError(
+                             "Both observation space and action space must be provided, if environment is None"
+            )
+    else:
+        if observation_space is None:
+            observation_space = env.observation_space
+        if action_space is None:
+            action_space = env.action_space
```
can you reduce the complexity of this making a single if for raising the error (then you don't need the else)

avjmachine replied:
Made the changes.

younik commented on minari/utils.py:
```diff
@@ -353,8 +353,8 @@ def get_average_reference_score(
 
 def create_dataset_from_buffers(
     dataset_id: str,
-    env: gym.Env,
     buffer: List[Dict[str, Union[list, Dict]]],
+    env: Optional[gym.Env] = None,
```
notice that this is a breaking change (positional calling will not work anymore), we may want to delay this for minari 0.5, that may be the next release anyway

avjmachine replied:
Okay, will work on the complexity (if conditions). 
Yes, it is a breaking change. I will keep the changes ready in any case.

I also had a question regarding the test case:

Which of the following options do we use for the env buffers in the absence of an env?
a. any buffers already available. I don't know if there is any, do you?
b. create buffer data by running one of the gym env
c. Manually create a dummy env buffer object

younik replied:
What do you mean with env buffers? We have a field in the metadata to save the env_spec; in case env is missing we will miss that attribute in the metadata. The env can missing, for example if the data comes from a real robot with no associated simulated environment. Does this answer your question?

avjmachine replied:
Let me rephrase my question. What do we pass as observation space and action space when env is None for the test case? Which one if those 3 options listed would be better?

younik replied:
If env is not specified, observation_space and action_space must be specified
This is what the if branch of lines 436-440 does, right?

younik replied:
The user must specify it

avjmachine replied:
Yes, the user must specify it.
That's why I was wondering what to use for the "test case". I wanted to check if we already have some action space observation space data for the test case, or should I create one. I listed those 3 options to get your advice on what would be the best option.

younik replied:
Oh clear now, you are asking it for tests
For that you can follow the same workflow of this: https://github.com/Farama-Foundation/Minari/blob/c43a61249ac51dd663a12983b7050aced97115b0/tests/utils/test_dataset_creation.py#L112

So creating the buffer from env (they are test env testing different spaces; I don't think it is necessary to test that function, one is enough). If you need to copy-paste the code from that test, feel free to create a method in tests/common.py with a function to create the buffer from an env

avjmachine replied:
Yes, thanks a lot, I was asking about the test. In the absence of env, we need to pass action space or observation space mandatorily. Hence, I wanted ideas on what arguments to pass for observation space and action space parameters when we test this scenario.

younik replied:
You should use the same observation_space and action_space of the environment that generates the dataset (`env.observation_space` and env.action_space`), then, for the sake of the test, you don't pass the environment but only these spaces

avjmachine replied:
Have resolved the complexity for if while raising error and suppressed pyright errors. Working on the test case.

avjmachine replied:
Have resolved the complexity for if while raising error and suppressed pyright errors. Working on the test case.

avjmachine commented on minari/utils.py:
```diff
@@ -482,18 +483,22 @@ def create_dataset_from_buffers(
-
-            file.attrs["action_space"] = action_space_str
-            file.attrs["observation_space"] = observation_space_str
+            if action_space is not None:
```
Changes are done

avjmachine commented on minari/utils.py:
```diff
@@ -433,10 +433,11 @@ def create_dataset_from_buffers(
-        observation_space = env.observation_space
-    if action_space is None:
-        action_space = env.action_space
+    if env is not None:
```
Changes are done

younik commented on minari/utils.py:
```diff
@@ -172,7 +172,8 @@ def validate_datasets_to_combine(datasets_to_combine: List[MinariDataset]) -> En
     else:
         max_episode_steps = max(
-            dataset.spec.env_spec.max_episode_steps for dataset in datasets_to_combine
+            dataset.spec.env_spec.max_episode_steps
+            for dataset in datasets_to_combine  # pyright: ignore[reportGeneralTypeIssues]
```
Actually, since now the datasets can have a env_spec=None, we should address also this case.
You can create the list of env_spec and filter out None values

avjmachine replied:
As of now, I'm raising an error if any of the datasets to be combined have no env_spec. Please do let me know if you'd prefer some other specific action instead of this.

younik commented on minari/utils.py:
```diff
@@ -292,7 +293,9 @@ def get_average_reference_score(
     for _ in range(num_episodes):
         while True:
             action = policy(obs)
-            obs, _, terminated, truncated, info = env.step(action)
+            obs, _, terminated, truncated, info = env.step(
+                action  # pyright: ignore[reportGeneralTypeIssues]
+            )
```
What was the warning here? I cannot see it

avjmachine replied:
It was the pyright raising error for reporting General Type Issue while running pre-commit hook.

younik commented on minari/utils.py:
```diff
@@ -394,10 +397,17 @@ def create_dataset_from_buffers(
     if observation_space is None:
-        observation_space = env.observation_space
+        observation_space = (
+            env.observation_space  # pyright: ignore[reportOptionalMemberAccess]
```
why do you need this? There is no way reaching here when env is None; do you got errors from the pre-commit?

avjmachine replied:
Yes, I got errors from pre-commit.

younik commented on minari/utils.py:
```diff
@@ -394,10 +397,17 @@ def create_dataset_from_buffers(
+        )
     if action_space is None:
-        action_space = env.action_space
+        action_space = env.action_space  # pyright: ignore[reportOptionalMemberAccess]
```
same as above

avjmachine replied:
Yes , pyright error for optional member access.

younik commented on tests/utils/test_dataset_combine.py:
```diff
@@ -72,8 +72,8 @@ def _generate_dataset_with_collector_env(
         assert env_spec is not None
-        env_spec.max_episode_steps = None
-        env = env_spec.make()
+        env_spec.max_episode_steps = None  # pyright: ignore[reportOptionalMemberAccess]
+        env = env_spec.make()  # pyright: ignore[reportOptionalMemberAccess]
```
env_space is not None for sure, I think you can remove the pyright ignore

avjmachine replied:
I've removed all pyright inline ignores. Please do let me know if I should set the pyright ignore in the .toml file or if there is some other option to resolve this.

younik reviewed (commented):
I had to play around a bit to understand how to solve so I just pushed the code to your branch; now should be good. 
Last thing before merging, can you add a test for combining datasets without env_spec?

avjmachine commented:
Okay, so using assert you could handle the pyright optional member access issue! Thanks a lot for helping with that!

younik commented on minari/utils.py:
```diff
@@ -161,31 +163,27 @@ def validate_datasets_to_combine(datasets_to_combine: List[MinariDataset]) -> En
+        env_spec = dataset.spec.env_spec
+        if env_spec is not None:
+            assert (
+                common_env_spec is not None
```
Oh my bad, you are right, thanks! Can you change this assert to check env_spec == common_env_spec? This should solve the issue

Would be nice to also add a test on combining with different env_spec then (an error is expected)

avjmachine replied:
okay, sure. I'll add the fix and the test as well.

younik replied:
Thanks for the fix! I left a couple of comments.
You are right about the assert, the difference is: when you believe something cannot happen you can use assert (e.g. asserts on line 401 and 404 of utils.py, as there is the check above). When it is something that can happen, but you want to let the user that it is not allowed, you should raise an error.
Unfortunately, in Minari we often use asserts as exceptions. I would be happy to merge another PR that addresses it.

> Edit: Just noticed that we need a separate for loop statement for updating the max_episode_steps after line 186. I missed that. Will correct it. But please do advise me on the assert statement.

What do you mean? I don't see it.

younik commented on minari/utils.py:
```diff
@@ -161,31 +163,27 @@ def validate_datasets_to_combine(datasets_to_combine: List[MinariDataset]) -> En
+            assert (
+                common_env_spec is not None
+            ), "Found incompatible env_spec in datasets"
+            if (
+                common_env_spec.max_episode_steps is None
+                or env_spec.max_episode_steps is None
+            ):
+                common_env_spec.max_episode_steps = None
+            else:
+                common_env_spec.max_episode_steps = max(
+                    common_env_spec.max_episode_steps, env_spec.max_episode_steps
+                )
```
this sounds good, but it still doesn't check if the two env_spec are the same.
One thing you can do is creating a copy of env_spec, modify the max_episode_steps to be the same of common_env_spec and then check env_spec == common_env_spec

avjmachine replied:
yes, that's correct. I'm doing the same in lines after 184. (the edit comment that I mentioned)

avjmachine replied:
Added the fix.

younik commented on tests/common.py:
```diff
@@ -403,6 +403,9 @@ def check_env_recovery_with_subset_spaces(
         observation_space_subset (gym.spaces.Space): desired subset observation space
 
     """
+    if dataset.env_spec is None:
+        raise ValueError("Recovering environment is not possible when env_spec is None")
```
this it is not necessary I believe, as `recover_environment` will throw an error itself if env_spec is None

avjmachine replied:
removed the redundant raise error

avjmachine commented:
@younik I'm currently working on the merge conflicts with the eval_env changes that got merged yesterday. I had one question regarding this. Could there be a scenario where there is an eval_env even when env is None? I mean, can a user provide a eval_env for a dataset generated with buffers without an env?

younik commented:
> @younik I'm currently working on the merge conflicts with the eval_env changes that got merged yesterday. I had one question regarding this. Could there be a scenario where there is an eval_env even when env is None? I mean, can a user provide a eval_env for a dataset generated with buffers without an env?

Thank you, good point. I don't see any common case when this would happen, but also no reason to constraint it. I would simply do what is easier in code, which is no constraint I guess.

avjmachine commented on minari/utils.py:
```diff
@@ -494,11 +508,24 @@ def create_dataset_from_buffers(
-        raise ValueError("The `env` argument must be of types str|EnvSpec|gym.Env")
+        raise ValueError("The `env` argument must be of types str|EnvSpec|gym.Env|None")
+
+    if isinstance(env, (str, EnvSpec, type(None))) and (observation_space is None or action_space is None):
+        raise ValueError("Both observation space and action space must be provided, if env is str|EnvSpec|None")
```
I'm assuming here that action_space and observation_space cannot be extracted unless it is of type gym.Env

avjmachine replied:
So, if it is a str or EnvSpec, there would be 2 steps to extract the action_space and observation_space? 

1. recover env as gym.Env with gym.make; 
2. recover action_space and observation_space from this gym.Env.

Is this correct?

younik replied:
Yes, exactly

avjmachine replied:
implemented the above logic

avjmachine commented on minari/utils.py:
```diff
@@ -516,18 +516,18 @@ def create_dataset_from_buffers(
-
+    if isinstance(env, (str, EnvSpec)):
+        env = gym.make(env)
+    assert env is not None
```
need this redundant statement to avoid pyright error

younik commented on minari/utils.py:
```diff
@@ -464,10 +485,10 @@ def create_dataset_from_buffers(
+        env (Optional[str|gym.Env|EnvSpec]): Gymnasium environment(gym.Env)/environment id(str)/environment spec(EnvSpec) used to collect the buffer data. Defaults to None.
         eval_env (Optional[str|gym.Env|EnvSpec]): Gymnasium environment(gym.Env)/environment id(str)/environment spec(EnvSpec) to use for evaluation with the dataset. After loading the dataset, the environment can be recovered as follows: `MinariDataset.recover_environment(eval_env=True).
-                                                If None the `env` used to collect the buffer data should be used for evaluation.
+                                                If None, and if the `env` used to collect the buffer data is available, latter should be used for evaluation.
```
not related to your PR, but `should` -> `will`

avjmachine replied:
updated

avjmachine commented on minari/utils.py:
```diff
@@ -161,31 +163,44 @@ def validate_datasets_to_combine(datasets_to_combine: List[MinariDataset]) -> En
-        combine_env_spec.append(dataset_env_spec)
+        assert isinstance(dataset, MinariDataset)
+        env_spec = dataset.spec.env_spec
+        if env_spec is not None:
```
okay, so if we have just 2 types of datasets - one type as None and other with an env_spec, then we throw an error? Or do we filter out the None types and just combine the others with a common env_spec?

younik replied:
We should throw an error

avjmachine replied:
changes implemented

avjmachine commented on minari/utils.py:
```diff
@@ -161,31 +163,44 @@ def validate_datasets_to_combine(datasets_to_combine: List[MinariDataset]) -> En
-    ), "The datasets to be combined have different values for `env_spec` attribute."
+    # checking equivalence of all datasets with an env
+    for dataset in datasets_to_combine:
+        if dataset.spec.env_spec is not None:
```
okay, I think this answers my question in the previous comment.

avjmachine replied:
updated

avjmachine commented on minari/utils.py:
```diff
@@ -392,11 +407,17 @@ def _generate_dataset_metadata(
+            warnings.warn(
+                "Since both `eval_env` and `env` used to collect the data are None, no environment can be used for evaluation",
+                UserWarning,
+            )
```
made the 2 checks independent

avjmachine commented on minari/utils.py:
```diff
@@ -161,31 +163,44 @@ def validate_datasets_to_combine(datasets_to_combine: List[MinariDataset]) -> En
+            if env_spec_copy != common_env_spec:
+                raise ValueError(
+                    "The datasets to be combined have different values for `env_spec` attribute."
+                )
```
done

younik reviewed (approved):
LGTM; thanks for implementing this!

younik commented:
Actually, can you fix pre-commit @avjmachine 

Looks like you should use `env_spec` in line 527 & 530 in utils, as env can be just a string.
For the documentation, it is not related to your PR, so you can ignore it

avjmachine commented:
> Actually, can you fix pre-commit @avjmachine
> 
> Looks like you should use `env_spec` in line 527 & 530 in utils, as env can be just a string. For the documentation, it is not related to your PR, so you can ignore it

In line 524, we are overwriting the env variable if its just a string, using "env = gym.make(env)". So, will this work or is there a risk of env not having an observation_space or action_space attribute even after using gym.make?

younik commented:
Yes, it is

> > Actually, can you fix pre-commit @avjmachine
> > Looks like you should use `env_spec` in line 527 & 530 in utils, as env can be just a string. For the documentation, it is not related to your PR, so you can ignore it
> 
> In line 524, we are overwriting the env variable if its just a string, using "env = gym.make(env)". So, will this work or is there a risk of env not having an observation_space or action_space attribute even after using gym.make?

No, there isn't; however, pyright is not able to infer it and throws an error.
Changing `assert env is not None` to `assert isinstance(env, gym.Env)` should work

avjmachine commented:
> Yes, it is
> 
> > > Actually, can you fix pre-commit @avjmachine
> > > Looks like you should use `env_spec` in line 527 & 530 in utils, as env can be just a string. For the documentation, it is not related to your PR, so you can ignore it
> > 
> > 
> > In line 524, we are overwriting the env variable if its just a string, using "env = gym.make(env)". So, will this work or is there a risk of env not having an observation_space or action_space attribute even after using gym.make?
> 
> No, there isn't; however, pyright is not able to infer it and throws an error. Changing `assert env is not None` to `assert isinstance(env, gym.Env)` should work

Done. I didn't get this pyright error on my local env. Hence, got confused.
Also, fixed additional pyright errors which I got while committing. You can review those changes. I changed int64 to integer for type check based on info [here](https://github.com/numpy/numpy/issues/2951) and [here](https://github.com/microsoft/pylance-release/issues/3830)
